rfqservice/stream.go

```diff
@@ -0,0 +1,239 @@
+package rfqservice
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightninglabs/taproot-assets/fn"
+	msg "github.com/lightninglabs/taproot-assets/rfqmessages"
+)
+
+// StreamHandler is a struct that handles incoming and outgoing peer RFQ stream
+// messages.
+//
+// This component subscribes to incoming raw peer messages (custom messages). It
+// processes those messages with the aim of extracting relevant request for
+// quotes (RFQs).
+type StreamHandler struct {
+	// peerMessagePorter is the peer message porter. This component is used
+	// to send and receive raw peer messages.
+	peerMessagePorter PeerMessagePorter
+
+	// recvRawMessages is a channel that receives incoming raw peer
+	// messages.
+	recvRawMessages <-chan lndclient.CustomMessage
+
+	// errRecvRawMessages is a channel that receives errors emanating from
+	// the peer raw messages subscription.
+	errRecvRawMessages <-chan error
+
+	// IncomingQuoteRequests is a channel which is populated with incoming
+	// (received) and valid quote request messages.
+	IncomingQuoteRequests *fn.EventReceiver[msg.QuoteRequest]
+
+	// errChan is this system's error reporting channel.
+	errChan chan error
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewStreamHandler creates and starts a new RFQ stream handler.
+func NewStreamHandler(ctx context.Context,
+	peerMsgPorter PeerMessagePorter) (*StreamHandler, error) {
+
+	msgChan, peerMsgErrChan, err := peerMsgPorter.SubscribeCustomMessages(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to subscribe to custom "+
+			"messages via message transfer handle: %w", err)
+	}
+
+	incomingQuoteRequests := fn.NewEventReceiver[msg.QuoteRequest](
+		fn.DefaultQueueSize,
+	)
+
+	streamHandler := StreamHandler{
+		peerMessagePorter:  peerMsgPorter,
+		recvRawMessages:    msgChan,
+		errRecvRawMessages: peerMsgErrChan,
+
+		IncomingQuoteRequests: incomingQuoteRequests,
+
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}
+
+	err = streamHandler.Start()
```

Comment:
Related to my comment above, but this will basically never return until shut down. So I think we should instead turn this into its own goroutine and start/stop management or call it `Run()` instead to make it clear it's going to block.